File: apps/dotnet/src/Core/BackupManager.cs
// Copyright 2026 Shazron Abdullah and Bunyi contributors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System.IO.Compression;
using Bunyi.Core.Diagnostics;
using Bunyi.Core.Models;

namespace Bunyi.Core;

/// <summary>How far a backup or restore has got (spec §6).</summary>
/// <param name="Fraction">0 to 1, or 0 when nothing is known yet.</param>
/// <param name="Detail">A line for a person, or null.</param>
public sealed record BackupProgress(double Fraction, string? Detail = null)
{
    /// <summary>The bytes-based fraction, as a percentage for a bar.</summary>
    public static BackupProgress Of(long done, long total, string? detail = null) =>
        new(total <= 0 ? 0 : Math.Clamp(done / (double)total, 0, 1), detail);
}

/// <summary>What a restore found in an archive, before touching anything.</summary>
/// <param name="Repos">The repositories it holds, as <c>org/name</c>.</param>
/// <param name="Bytes">How much will be written if all of them are.</param>
public sealed record BackupContents(IReadOnlyList<string> Repos, long Bytes);

/// <summary>
/// Backing the models folder up, and putting it back (spec §6).
/// </summary>
/// <remarks>
/// <para>
/// The point of this is not tidiness: a models folder is several gigabytes
/// fetched over a slow link, and the alternative to a backup is downloading it
/// again on the next machine.
/// </para>
/// <para>
/// <b>Stored, never compressed.</b> Weights are already incompressible, so
/// deflating them spends a great deal of CPU to save almost nothing — and it
/// costs the determinate progress bar, because a compressed archive's size no
/// longer tracks the bytes read.
/// </para>
/// </remarks>
public sealed class BackupManager(ILogSink log)
{
    private readonly ILogSink _log = log ?? throw new ArgumentNullException(nameof(log));

    /// <summary>The folder inside an archive that holds the repositories.</summary>
    internal const string ModelsEntry = "models";

    /// <summary>
    /// Archives the models folder to one zip (spec §6).
    /// </summary>
    /// <remarks>
    /// Built beside its destination under a temporary name and moved into place
    /// at the end. The move is a rename within one directory, so it cannot half
    /// happen — and a run that is cancelled or fails leaves nothing at the path
    /// the user chose, rather than a truncated archive that looks like a backup.
    /// </remarks>
    public async Task BackupAsync(
        string modelsFolder,
        string destinationZip,
        IProgress<BackupProgress>? progress,
        CancellationToken ct)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(modelsFolder);
        ArgumentException.ThrowIfNullOrWhiteSpace(destinationZip);

        if (!Directory.Exists(modelsFolder))
        {
            throw new DirectoryNotFoundException(
                $"There is no models folder at {modelsFolder} to back up.");
        }

        var files = Directory
            .EnumerateFiles(modelsFolder, "*", SearchOption.AllDirectories)
            .Where(f => !f.EndsWith(".incomplete", StringComparison.OrdinalIgnoreCase))
            .ToList();

        if (files.Count == 0)
        {
            throw new InvalidOperationException(
                "There are no models to back up yet. Generate something first, or "
                + "download a model in Settings.");
        }

        var total = files.Sum(f => new FileInfo(f).Length);
        _log.Log($"Backing up {files.Count} files ({Bytes(total)}) to {destinationZip}.");

        var folder = Path.GetDirectoryName(Path.GetFullPath(destinationZip));
        if (!string.IsNullOrEmpty(folder)) Directory.CreateDirectory(folder);

        var temporary = destinationZip + ".partial";

        try
        {
            var done = 0L;

            await using (var stream = File.Create(temporary))
            using (var archive = new ZipArchive(stream, ZipArchiveMode.Create))
            {
                foreach (var file in files)
                {
                    ct.ThrowIfCancellationRequested();

                    var relative = Path.GetRelativePath(modelsFolder, file)
                        .Replace(Path.DirectorySeparatorChar, '/');

                    // NoCompression: see the class remarks. This is the whole
                    // reason a backup of 4 GB takes a minute rather than ten.
                    var entry = archive.CreateEntry(relative, CompressionLevel.NoCompression);

                    await using var source = File.OpenRead(file);
                    await using var target = entry.Open();
                    done = await CopyAsync(source, target, done, total, relative, progress, ct)
                        .ConfigureAwait(false);
                }
            }

            File.Move(temporary, destinationZip, overwrite: true);
            progress?.Report(new BackupProgress(1, "Finished."));

            _log.Log($"Backed up to {destinationZip} ({Bytes(new FileInfo(destinationZip).Length)}).");
        }
        catch
        {
            // Nothing half-written survives, at the temporary name or the real
            // one. A cancelled backup should leave the folder as it found it.
            Delete(temporary);
            throw;
        }
    }

    /// <summary>
    /// Reads what an archive holds, without writing anything (spec §6).
    /// </summary>
    /// <remarks>
    /// Separate from the restore so the window can say what is about to happen
    /// — and so an archive that is not a Bunyi backup is refused before any
    /// file is touched, rather than half way through.
    /// </remarks>
    public BackupContents Inspect(string sourceZip)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(sourceZip);

        if (!File.Exists(sourceZip))
        {
            throw new FileNotFoundException($"There is no backup at {sourceZip}.", sourceZip);
        }

        using var archive = ZipFile.OpenRead(sourceZip);

        var prefix = FindModels(archive)
            ?? throw new InvalidDataException(
                "That zip does not look like a Bunyi backup — it has no models folder "
                + "inside it. Choose the .zip a backup produced.");

        var repos = new SortedSet<string>(StringComparer.OrdinalIgnoreCase);
        var bytes = 0L;

        foreach (var entry in archive.Entries)
        {
            if (RepoOf(entry.FullName, prefix) is not { } repo) continue;

            repos.Add(repo);
            bytes += entry.Length;
        }

        if (repos.Count == 0)
        {
            throw new InvalidDataException(
                "That backup has a models folder but no models in it.");
        }

        return new BackupContents([.. repos], bytes);
    }

    /// <summary>
    /// Merges an archive into the models folder (spec §6).
    /// </summary>
    /// <remarks>
    /// <b>Never clobbers.</b> A repository already on disk is skipped whole
    /// rather than merged file by file: a half-replaced model is the one
    /// outcome worse than either keeping or replacing it, and the folder on
    /// disk is the one that has already been verified complete.
    /// </remarks>
    public async Task<IReadOnlyList<string>> RestoreAsync(
        string sourceZip,
        string modelsFolder,
        IProgress<BackupProgress>? progress,
        CancellationToken ct)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(modelsFolder);

        var contents = Inspect(sourceZip);

        using var archive = ZipFile.OpenRead(sourceZip);
        var prefix = FindModels(archive)!;

        var skipped = new List<string>();
        var wanted = new HashSet<string>(StringComparer.OrdinalIgnoreCase);

        foreach (var repo in contents.Repos)
        {
            var already = Path.Combine(
                modelsFolder, ModelsEntry, repo.Replace('/', Path.DirectorySeparatorChar));

            if (Directory.Exists(already) && Directory.EnumerateFileSystemEntries(already).Any())
            {
                skipped.Add(repo);
                _log.Log($"Keeping the copy of {repo} already here; the backup's is skipped.");
            }
            else
            {
                wanted.Add(repo);
            }
        }

        var total = archive.Entries
            .Where(e => RepoOf(e.FullName, prefix) is { } r && wanted.Contains(r))
            .Sum(e => e.Length);

        _log.Log(
            $"Restoring {wanted.Count} of {contents.Repos.Count} model(s) ({Bytes(total)}); "
            + $"{skipped.Count} already here.");

        var done = 0L;
        var restored = new List<string>();

        foreach (var entry in archive.Entries)
        {
            ct.ThrowIfCancellationRequested();

            if (RepoOf(entry.FullName, prefix) is not { } repo || !wanted.Contains(repo)) continue;

            // The archive's own paths decide where files land, so they are
            // checked the same way a download manifest's are: an entry that
            // climbs out of the models folder is refused, not written.
            var relative = entry.FullName[prefix.Length..].TrimStart('/');
            if (!ManifestPath.TryNormalize(relative, out var safe))
            {
                _log.Log($"Skipped an unsafe path in the backup: {entry.FullName}");
                continue;
            }

            var destination = Path.Combine(modelsFolder, ModelsEntry, safe);
            Directory.CreateDirectory(Path.GetDirectoryName(destination)!);

            await using (var source = entry.Open())
            await using (var target = File.Create(destination))
            {
                done = await CopyAsync(source, target, done, total, repo, progress, ct)
                    .ConfigureAwait(false);
            }

            if (!restored.Contains(repo)) restored.Add(repo);
        }

        progress?.Report(new BackupProgress(1, "Finished."));
        _log.Log($"Restored {restored.Count} model(s).");

        return skipped;
    }

    /// <summary>
    /// The <c>models/</c> prefix inside an archive, or null if there is none.
    /// </summary>
    /// <remarks>
    /// "At or near the root": a zip made by a file manager often wraps
    /// everything in a folder named after the archive, and refusing those would
    /// reject backups that are perfectly good.
    /// </remarks>
    internal static string? FindModels(ZipArchive archive)
    {
        ArgumentNullException.ThrowIfNull(archive);

        foreach (var entry in archive.Entries)
        {
            var parts = entry.FullName.Split('/');

            for (var i = 0; i < Math.Min(parts.Length, 2); i++)
            {
                if (string.Equals(parts[i], ModelsEntry, StringComparison.OrdinalIgnoreCase))
                {
                    return string.Join('/', parts[..(i + 1)]) + "/";
                }
            }
        }

        return null;
    }

    /// <summary>The <c>org/name</c> an entry belongs to, or null.</summary>
    internal static string? RepoOf(string entryPath, string prefix)
    {
        if (!entryPath.StartsWith(prefix, StringComparison.OrdinalIgnoreCase)) return null;

        var rest = entryPath[prefix.Length..].Split('/');

        // org/name/file at the least. A file sitting directly in models/
        // belongs to no repository and is not restored.
        return rest.Length >= 3 ? $"{rest[0]}/{rest[1]}" : null;
    }

    /// <returns>The running total, since an async method cannot take it by ref.</returns>
    private static async Task<long> CopyAsync(
        Stream source, Stream target,
        long done, long total, string what,
        IProgress<BackupProgress>? progress, CancellationToken ct)
    {
        var buffer = new byte[81_920];
        var sinceReport = 0L;
        var running = done;

        int read;
        while ((read = await source.ReadAsync(buffer, ct).ConfigureAwait(false)) > 0)
        {
            await target.WriteAsync(buffer.AsMemory(0, read), ct).ConfigureAwait(false);

            running += read;
            sinceReport += read;

            // Every few megabytes rather than every buffer: a 4 GB backup is
            // fifty thousand buffers, and reporting each one costs more than the
            // copying does.
            if (sinceReport >= 4_000_000)
            {
                sinceReport = 0;
                progress?.Report(BackupProgress.Of(running, total, what));
            }
        }

        return running;
    }

    private void Delete(string path)
    {
        try
        {
            if (File.Exists(path)) File.Delete(path);
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
        {
            _log.Log($"Could not clean up {path}. {ex.Message}");
        }
    }

    private static string Bytes(long bytes) => bytes switch
    {
        >= 1_000_000_000 => $"{bytes / 1_000_000_000.0:0.0} GB",
        >= 1_000_000 => $"{bytes / 1_000_000.0:0} MB",
        _ => $"{bytes / 1_000.0:0} KB",
    };
}

File: apps/macos/BackupManager.swift
// Copyright 2026 Shazron Abdullah and Bunyi contributors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//
//  BackupManager.swift
//  Bunyi
//
//  Backs up the models folder to a single zip and restores from one.
//  Sandbox shapes both directions. The default container folder is zipped
//  with `zip -0` (stored, no compression) so it's fast — model weights are
//  already-incompressible safetensors — and the output size tracks the
//  source, which drives a real progress bar. A custom (bookmarked) folder
//  isn't reachable by a child process, so it falls back to NSFileCoordinator
//  (in-process, no progress). Restore extracts with /usr/bin/ditto on
//  container-local temp paths, then merges in-process.
//

import Foundation

@MainActor
@Observable
final class BackupManager {
    enum Status: Equatable {
        case idle
        case working(String)
        /// Stop was pressed and the child process is being torn down. A
        /// distinct state because the work does not end the instant Stop is
        /// pressed: `zip` has to die, a partial archive has to be removed, and
        /// on a 23 GB backup that gap was long enough that nothing appeared to
        /// happen — so Stop got pressed again. The engine already models this
        /// for generation (`EngineStatus.stopping`); backup did not.
        case stopping
        case done(String)
        case error(String)

        var isBusy: Bool {
            switch self {
            case .working, .stopping: true
            default: false
            }
        }

        /// Whether Stop can still do anything. False once stopping, so the
        /// button can be disabled rather than inviting a second press that
        /// only logs a second "Stopping…".
        var canCancel: Bool {
            if case .working = self { return true }
            return false
        }
    }

    var status: Status = .idle
    /// 0...1 while zipping when a total size is known; nil = indeterminate.
    var progress: Double?

    private let log = LogStore.shared
    private var currentTask: Task<Void, Never>?
    /// The running progress poller, held so `cancel()` can stop it reporting.
    private var sizeMonitor: Task<Void, Never>?
    /// Lets `cancel()` (main actor) terminate the child zip/ditto process
    /// created inside a detached task.
    private let running = RunningProcess()

    func cancel() {
        guard status.canCancel else { return }
        log.log("Stopping…")
        status = .stopping
        // The size monitor polls a directory the child process is in the
        // middle of abandoning, so left running it reports the archive
        // shrinking — "Archiving 0% — Zero kB" moments after Stop, which reads
        // as the backup starting over rather than ending. It is an unstructured
        // Task, so cancelling `currentTask` does not reach it; it has to be
        // cancelled by hand.
        sizeMonitor?.cancel()
        sizeMonitor = nil
        progress = nil
        running.cancel()
        currentTask?.cancel()
    }

    // MARK: Backup

    func startBackup(to destination: URL) {
        currentTask?.cancel()
        running.reset()
        currentTask = Task { await self.runBackup(to: destination) }
    }

    private func runBackup(to destination: URL) async {
        let source = ModelsLocation.current()
        let modelsSubdir = source.appendingPathComponent("models", isDirectory: true)
        let fastPath = !ModelsLocation.isCustom
        // zip builds into its own working file (ziXXXXXX) in the output's
        // directory and only renames to the final name at the end, so give
        // it a dedicated dir and monitor the whole dir, not the final name.
        let tempDir = FileManager.default.temporaryDirectory
            .appendingPathComponent("backup-\(UUID().uuidString)", isDirectory: true)
        let tempZip = tempDir.appendingPathComponent("models.zip")

        status = .working("Creating archive…")
        progress = nil
        log.log("Backing up models from \(source.path)")

        do {
            if fastPath {
                try FileManager.default.createDirectory(
                    at: tempDir, withIntermediateDirectories: true)
                let total = Self.directorySize(modelsSubdir)
                log.log("Creating \(destination.lastPathComponent) — "
                    + "\(total.formatted(.byteCount(style: .file))) to archive "
                    + "(stored, no compression)")
                let monitor = startSizeMonitor(total: total, label: "Archiving") {
                    Self.directorySize(tempDir)
                }
                do {
                    try await Task.detached(priority: .utility) { [running] in
                        try Self.createZipStored(
                            sourceFolder: source, to: tempZip, control: running)
                    }.value
                } catch {
                    monitor.cancel()
                    throw error
                }
                monitor.cancel()
                try Task.checkCancellation()

                // The child zip built inside the container (it can't write to
                // the sandbox-granted destination directly). Move it out — but
                // OFF the main thread, or a slow save freezes the UI. On the
                // same volume that's an instant rename; on a different volume
                // (network/external drive) it's a real multi-GB copy, so stream
                // it with a progress bar and Stop support.
                status = .working("Saving…")
                progress = nil
                if Self.sameVolume(tempZip, destination) {
                    try await Task.detached(priority: .utility) {
                        let fm = FileManager.default
                        if fm.fileExists(atPath: destination.path) {
                            try fm.removeItem(at: destination)
                        }
                        try fm.moveItem(at: tempZip, to: destination)
                    }.value
                } else {
                    let bytes = Self.fileSize(tempZip)
                    log.log("Saving to another drive — copying "
                        + "\(bytes.formatted(.byteCount(style: .file)))")
                    let saveMonitor = startSizeMonitor(total: bytes, label: "Saving") {
                        Self.fileSize(destination)
                    }
                    do {
                        try await Task.detached(priority: .utility) { [running] in
                            try Self.copyFile(
                                from: tempZip, to: destination, control: running)
                        }.value
                    } catch {
                        saveMonitor.cancel()
                        throw error
                    }
                    saveMonitor.cancel()
                }
            } else {
                log.log("Creating \(destination.lastPathComponent) — this can "
                    + "take a few minutes (custom folder: no progress, and "
                    + "Stop can't interrupt archiving mid-file)")
                try await Task.detached(priority: .utility) {
                    try Self.createZipCoordinated(of: source, to: destination)
                }.value
            }

            try? FileManager.default.removeItem(at: tempDir)
            progress = nil
            let size = (try? destination.resourceValues(forKeys: [.fileSizeKey])
                .fileSize).map { Int64($0).formatted(.byteCount(style: .file)) }
                ?? "unknown size"
            log.log("Backup finished: \(destination.path) (\(size))")
            status = .done("Backed up \(size) to \(destination.lastPathComponent)")
        } catch is CancellationError {
            try? FileManager.default.removeItem(at: tempDir)
            if fastPath { try? FileManager.default.removeItem(at: destination) }
            progress = nil
            log.log("Backup cancelled")
            status = .idle
        } catch {
            try? FileManager.default.removeItem(at: tempDir)
            progress = nil
            log.log("Backup failed: \(String(describing: error))")
            status = .error(error.localizedDescription)
        }
    }

    /// Polls a growing output so a determinate bar can show real progress —
    /// the zip's working dir while archiving, or the destination file while
    /// copying to another volume. `measure` returns bytes written so far.
    /// Starts a poller, and records it as the one `cancel()` should stop.
    ///
    /// Recorded here rather than by the caller, because leaving it to callers
    /// is exactly what went wrong: only the archiving monitor was stored, so
    /// pressing Stop during the cross-volume "Saving…" copy left that monitor
    /// running — still logging, and still writing `progress` after `cancel()`
    /// had cleared it. Doing it here means any monitor added later is tracked
    /// without anyone remembering to.
    private func startSizeMonitor(
        total: Int64, label: String, measure: @escaping @Sendable () -> Int64
    ) -> Task<Void, Never> {
        let task = Task { [weak self, log] in
            guard total > 0 else { return }
            var lastDecile = -1
            while !Task.isCancelled {
                let size = measure()
                let fraction = min(Double(size) / Double(total), 1.0)
                self?.progress = fraction
                let decile = Int(fraction * 10)
                if decile != lastDecile {
                    lastDecile = decile
                    log.log("\(label) \(Int(fraction * 100))% — "
                        + "\(size.formatted(.byteCount(style: .file)))")
                }
                try? await Task.sleep(for: .seconds(1))
            }
        }
        sizeMonitor = task
        return task
    }

    nonisolated private static func createZipStored(
        sourceFolder: URL, to zip: URL, control: RunningProcess
    ) throws {
        let errorPipe = Pipe()
        let process = try control.start {
            let p = Process()
            p.executableURL = URL(fileURLWithPath: "/usr/bin/zip")
            p.currentDirectoryURL = sourceFolder
            // -0 store, -r recurse, -q quiet, -X drop extra attrs.
            p.arguments = ["-0", "-r", "-q", "-X", zip.path, "models"]
            p.standardError = errorPipe
            return p
        }
        try process.run()
        process.waitUntilExit()
        control.finish()
        guard process.terminationStatus == 0 else {
            if control.wasCancelled { throw CancellationError() }
            let message = String(
                data: errorPipe.fileHandleForReading.readDataToEndOfFile(),
                encoding: .utf8) ?? ""
            throw BackupError.zipFailed(message)
        }
    }

    /// NSFileCoordinator with .forUploading yields a zipped temp copy of
    /// the directory; the copy must happen inside the accessor block.
    nonisolated private static func createZipCoordinated(
        of source: URL, to destination: URL
    ) throws {
        let coordinator = NSFileCoordinator()
        var coordinationError: NSError?
        var copyError: Error?
        coordinator.coordinate(readingItemAt: source, options: .forUploading,
                               error: &coordinationError) { zipURL in
            do {
                let fm = FileManager.default
                if fm.fileExists(atPath: destination.path) {
                    try fm.removeItem(at: destination)
                }
                try fm.copyItem(at: zipURL, to: destination)
            } catch {
                copyError = error
            }
        }
        if let coordinationError { throw coordinationError }
        if let copyError { throw copyError }
    }

    // MARK: Restore

    func startRestore(from zipURL: URL) {
        currentTask?.cancel()
        running.reset()
        currentTask = Task { await self.runRestore(from: zipURL) }
    }

    private func runRestore(from zipURL: URL) async {
        let target = ModelsLocation.current()
        status = .working("Restoring…")
        progress = nil
        log.log("Restoring models from \(zipURL.lastPathComponent)")
        do {
            let result = try await Task.detached(priority: .utility) { [running] in
                try Self.performRestore(zip: zipURL, into: target, control: running)
            }.value
            for name in result.restored { log.log("Restored \(name)") }
            for name in result.skipped { log.log("Skipped \(name) (already present)") }
            let summary = result.restored.isEmpty
                ? "Nothing to restore — all models in the backup are already present"
                : "Restored \(result.restored.count) model\(result.restored.count == 1 ? "" : "s")"
            log.log(summary)
            status = .done(summary)
        } catch is CancellationError {
            log.log("Restore cancelled")
            status = .idle
        } catch {
            log.log("Restore failed: \(String(describing: error))")
            status = .error(error.localizedDescription)
        }
    }

    private struct RestoreResult: Sendable {
        var restored: [String] = []
        var skipped: [String] = []
    }

    nonisolated private static func performRestore(
        zip: URL, into target: URL, control: RunningProcess
    ) throws -> RestoreResult {
        let fm = FileManager.default
        let tmp = fm.temporaryDirectory
            .appendingPathComponent("restore-\(UUID().uuidString)", isDirectory: true)
        defer { try? fm.removeItem(at: tmp) }
        let extractDir = tmp.appendingPathComponent("extract", isDirectory: true)
        try fm.createDirectory(at: extractDir, withIntermediateDirectories: true)

        // Copy the zip into the container so ditto (which only inherits the
        // plain sandbox, not the open-panel grant) can read it.
        let localZip = tmp.appendingPathComponent("backup.zip")
        let gotAccess = zip.startAccessingSecurityScopedResource()
        defer { if gotAccess { zip.stopAccessingSecurityScopedResource() } }
        try fm.copyItem(at: zip, to: localZip)
        try extractZip(localZip, to: extractDir, control: control)

        guard let modelsRoot = findModelsRoot(in: extractDir) else {
            throw BackupError.notAModelsBackup
        }

        let targetModels = target.appendingPathComponent("models", isDirectory: true)
        var result = RestoreResult()
        for org in try subdirectories(of: modelsRoot) {
            try Task.checkCancellation()
            let targetOrg = targetModels
                .appendingPathComponent(org.lastPathComponent, isDirectory: true)
            try fm.createDirectory(at: targetOrg, withIntermediateDirectories: true)
            for repo in try subdirectories(of: org) {
                let name = "\(org.lastPathComponent)/\(repo.lastPathComponent)"
                let dest = targetOrg
                    .appendingPathComponent(repo.lastPathComponent, isDirectory: true)
                if fm.fileExists(atPath: dest.path) {
                    result.skipped.append(name)
                } else {
                    try fm.moveItem(at: repo, to: dest)
                    result.restored.append(name)
                }
            }
        }
        return result
    }

    nonisolated private static func extractZip(
        _ zip: URL, to dir: URL, control: RunningProcess
    ) throws {
        let errorPipe = Pipe()
        let process = try control.start {
            let p = Process()
            p.executableURL = URL(fileURLWithPath: "/usr/bin/ditto")
            p.arguments = ["-xk", zip.path, dir.path]
            p.standardError = errorPipe
            return p
        }
        try process.run()
        process.waitUntilExit()
        control.finish()
        guard process.terminationStatus == 0 else {
            if control.wasCancelled { throw CancellationError() }
            let message = String(
                data: errorPipe.fileHandleForReading.readDataToEndOfFile(),
                encoding: .utf8) ?? ""
            throw BackupError.extractFailed(message)
        }
    }

    /// The zip's root varies ("models/…" if the user zipped that level,
    /// "Models/models/…" from our own backups), so find the shallowest
    /// directory literally named "models".
    nonisolated private static func findModelsRoot(in dir: URL) -> URL? {
        var queue: [(URL, Int)] = [(dir, 0)]
        while !queue.isEmpty {
            let (url, depth) = queue.removeFirst()
            if depth > 0, url.lastPathComponent == "models" { return url }
            guard depth < 3 else { continue }
            for sub in (try? subdirectories(of: url)) ?? [] {
                queue.append((sub, depth + 1))
            }
        }
        return nil
    }

    nonisolated private static func subdirectories(of dir: URL) throws -> [URL] {
        try FileManager.default.contentsOfDirectory(
            at: dir, includingPropertiesForKeys: [.isDirectoryKey],
            options: [.skipsHiddenFiles]
        ).filter { (try? $0.resourceValues(forKeys: [.isDirectoryKey]).isDirectory) == true }
    }

    // MARK: Sizes

    nonisolated private static func directorySize(_ dir: URL) -> Int64 {
        var total: Int64 = 0
        let keys: [URLResourceKey] = [.fileSizeKey, .totalFileAllocatedSizeKey]
        if let e = FileManager.default.enumerator(
            at: dir, includingPropertiesForKeys: keys) {
            for case let url as URL in e {
                let v = try? url.resourceValues(forKeys: Set(keys))
                total += Int64(v?.fileSize ?? v?.totalFileAllocatedSize ?? 0)
            }
        }
        return total
    }

    nonisolated private static func fileSize(_ url: URL) -> Int64 {
        Int64((try? url.resourceValues(forKeys: [.fileSizeKey]).fileSize) ?? 0)
    }

    /// Whether both paths live on the same volume, so a move is a rename.
    /// `b` may not exist yet, so its parent directory is checked. Unknown →
    /// false, which routes through the streamed copy (correct either way).
    nonisolated private static func sameVolume(_ a: URL, _ b: URL) -> Bool {
        let va = try? a.resourceValues(forKeys: [.volumeIdentifierKey])
            .volumeIdentifier
        let vb = try? b.deletingLastPathComponent()
            .resourceValues(forKeys: [.volumeIdentifierKey]).volumeIdentifier
        guard let va, let vb else { return false }
        return va.isEqual(vb)
    }

    /// Streamed copy so a slow cross-volume save (network/external drive) runs
    /// off the main thread with progress and honors Stop. Writes straight to
    /// the destination path, which the size monitor polls.
    nonisolated private static func copyFile(
        from src: URL, to dst: URL, control: RunningProcess
    ) throws {
        let fm = FileManager.default
        if fm.fileExists(atPath: dst.path) { try fm.removeItem(at: dst) }
        guard fm.createFile(atPath: dst.path, contents: nil) else {
            throw CocoaError(.fileWriteUnknown)
        }
        let input = try FileHandle(forReadingFrom: src)
        defer { try? input.close() }
        let output = try FileHandle(forWritingTo: dst)
        defer { try? output.close() }
        let chunk = 8 * 1024 * 1024
        while true {
            if control.wasCancelled {
                try? output.close()
                try? fm.removeItem(at: dst)
                throw CancellationError()
            }
            let data = try input.read(upToCount: chunk) ?? Data()
            if data.isEmpty { break }
            try output.write(contentsOf: data)
        }
    }
}

/// Thread-safe holder so a cancel on the main actor can terminate a child
/// process started inside a detached task.
final class RunningProcess: @unchecked Sendable {
    private let lock = NSLock()
    private var process: Process?
    private(set) var wasCancelled = false

    func reset() {
        lock.lock(); defer { lock.unlock() }
        process = nil
        wasCancelled = false
    }

    func start(_ make: () -> Process) throws -> Process {
        lock.lock(); defer { lock.unlock() }
        if wasCancelled { throw CancellationError() }
        let p = make()
        process = p
        return p
    }

    func finish() {
        lock.lock(); defer { lock.unlock() }
        process = nil
    }

    func cancel() {
        lock.lock()
        wasCancelled = true
        let p = process
        lock.unlock()
        p?.terminate()
    }
}

enum BackupError: LocalizedError {
    case notAModelsBackup
    case zipFailed(String)
    case extractFailed(String)

    var errorDescription: String? {
        switch self {
        case .notAModelsBackup:
            "That zip doesn't look like a models backup — no models folder inside."
        case .zipFailed(let detail):
            "Couldn't create the archive. \(detail)"
        case .extractFailed(let detail):
            "Couldn't unpack the backup. \(detail)"
        }
    }
}
